paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+                return volume
+        log.warning('Could not determine total simulation volume.')
+        return None
+
+    def remove_waters_manually(self):
+        """
+        Remove a few water molecules manually with `tleap` to exactly match a desired number of waters.
+
+        """
+    
+        cycle = 0
+        max_cycles = 10
+        waters = self.wat_added_history[-1]
+        while waters > self.target_waters:
+            # Retrieve excess water residues
+            water_surplus = (waters - self.target_waters)
+            water_residues = self.list_waters()
+            self.waters_to_remove = water_residues[-1 * water_surplus:]
+            log.debug('Manually removing waters... {}'.format(self.waters_to_remove))
+
+            # Get counts for all residues
+            residues = self.count_residues()
+
+            # Check if we reached target
+            waters = residues['WAT']
+            if waters == self.target_waters:
+                for key, value in sorted(residues.items()):
+                    log.info('{}\t{}'.format(key, value))
+                return
+            cycle += 1
+            if cycle > max_cycles:
+                raise Exception(
+                    "Solvation failed due to an unanticipated problem with water removal."
+                )
+
+    def list_waters(self):
+        """
+        Run and parse `tleap` output and return the a list of water residues.
+    
+        Returns
+        -------
+        water_residues : list
+            A list of the water residues in the structure
+
+        """
+        output = self.run()
+    
+        # Return a list of residue numbers for the waters
+        water_residues = []
+        for line in output[0].splitlines():
+            # Is this line a water?
+            match = re.search("^R<WAT (.*)>", line)
+            if match:
+                water_residues.append(match.group(1))
+        return water_residues
+
+    def adjust_buffer_value(self):
+        """
+        Determine whether to increase or decrease the buffer thickness to match a desired number of waters.
+    
+        Sets
+        -------
+        self.buffer_value : float
+            A new buffer size to try
+        self.exponent : int
+            Adjusts the order of magnitue of buffer value changes
+        self.cyc_since_last_exp_change : int
+            Resets this value to 0 when exponent to help with the logic gates.
+
+        """
+    
+        # If the number of waters was less than the target and is now greater than the target, make the buffer a bit
+        # smaller, by an increasingly smaller amount, (if we've taken more than one step since the last exponent change)
+        if self.wat_added_history[-2] < self.target_waters and self.wat_added_history[-1] > self.target_waters and self.cyc_since_last_exp_change > 1:
+            log.debug('Adjustment loop 1')
+            self.exponent -= 1
+            self.cyc_since_last_exp_change = 0
```

Comment:
Is keeping track of this what makes this an improvement over the previous version of solvate?

Reply:
Yes, that creates two extra if/elif statements now which means that we don't jump to tiny buffer_value increments too fast.